### PR TITLE
BC17-2,3: Implement Notification UI and Integration API

### PR DESCRIPTION
**Pull request by NguyenDoQuangMinh2112**
Build the Notification UI
- Dropdown UI
- Message notification 

Intergration API Notification
- Application web socket for realtime.

### CODE REVIEW

File: src/components/Layouts/components/ChatInterface/CommentField/CommentField.tsx [hunk 1]
```diff
@@ -62,32 +64,10 @@ const CommentField = ({
       }
       const payload = {
         blog_id: id,
-        blog_author: userInfo._id,
+        blog_author: blogDetail?.author._id,
```
Comment (nash-nguyen) [inline src/components/Layouts/components/ChatInterface/CommentField/CommentField.tsx]: check null for _id
`blog_author: blogDetail?.author?._id,`

File: src/components/Button/Button.module.scss [hunk 2]
```diff
@@ -215,3 +215,22 @@
     color: white !important;
   }
 }
+//
```
Comment (nash-nguyen) [inline src/components/Button/Button.module.scss]: remove empty comment

File: src/apis/notification/index.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import authorizedAxiosInstance from '~/axios'
+import { ApiResponse } from '~/interfaces/ApiResponse'
+import { NotificationInterface } from '~/interfaces/notification'
+import { API_ROOT } from '~/utils/constant'
+
+export const getNotificationByUserIdAPI = async (): Promise<ApiResponse<NotificationInterface[]>> => {
+  return await authorizedAxiosInstance.get(`${API_ROOT}/v1/notification`)
+}
```
Comment (nash-nguyen) [inline src/apis/notification/index.ts]: `export const getNotificationByUserIdAPI = (): Promise<ApiResponse<NotificationInterface[]>> => 
  authorizedAxiosInstance.get(`${API_ROOT}/v1/notification`);`
  
  It no need to return await in the async function

File: src/components/Layouts/components/Notification/ItemNotification.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import moment from 'moment'
+
+const cx = classNames.bind(styles)
+
+interface ItemNotificationInterface {
+  data: NotificationInterface
+}
+const ItemNotification = ({ data }: ItemNotificationInterface) => {
+  return (
+    <div className={cx('notification', { no_seen: !data?.seen })}>
```
Comment (nash-nguyen) [inline src/components/Layouts/components/Notification/ItemNotification.tsx:13]: Does call cx() func everytime in the component affect to the performance? If component has effect trigger other render, the cx() will be call again

File: src/components/Layouts/components/Notification/Notification.tsx [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+import { useSelector } from 'react-redux'
+import { authSelector } from '~/redux/auth/authSelectors'
+const cx = classNames.bind(styles)
+
+const Notification = () => {
+  const [notifications, setNotifications] = useState<NotificationInterface[] | null>(null)
+  const { userInfo } = useSelector(authSelector)
+  const { socket } = useSelector((state: any) => state.socket)
+
+  const fetchNotificationsApi = async () => {
```
Comment (nash-nguyen) [inline src/components/Layouts/components/Notification/Notification.tsx:22]: Suggest to use the `useCallback` whenever init a function in reactJS

File: src/components/Layouts/components/Notification/Notification.tsx [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+          return [notification, ...(prev || [])]
+        })
+      })
+    }
+    return () => {
+      socket?.off('newNotification')
+    }
+  }, [socket])
+
+  const markAllAsRead = () => {
```
Comment (nash-nguyen) [inline src/components/Layouts/components/Notification/Notification.tsx]: Suggest to  use the `useCallback` hook and move this function to the top, same level with the `fetchNotificationsApi`

File: src/components/Layouts/components/Notification/Notification.tsx (hunk 5 above)
Comment (nash-nguyen) [inline src/components/Layouts/components/Notification/Notification.tsx]: Suggest to use the `useCallback` hook

File: src/components/Layouts/components/Notification/Notification.tsx [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  useEffect(() => {
+    fetchNotificationsApi()
+
+    return () => {
+      socket?.off('newNotification')
+    }
+  }, [socket])
+  useEffect(() => {
```
Comment (nash-nguyen) [inline src/components/Layouts/components/Notification/Notification.tsx:51]: add a empty line above to seperate the effect -> will have more readability

File: src/components/Layouts/components/Search/Search.tsx [hunk 8]
```diff
@@ -95,7 +94,7 @@ const Search = () => {
                       <h5 className={cx('search-title')}>Bài viết</h5>
                       <span>Xem thêm</span>
                     </div>
-                    {searchResult.blogs.map((blog, index) => (
+                    {searchResult.blogs.map((blog) => (
```
Comment (nash-nguyen) [inline src/components/Layouts/components/Search/Search.tsx]: `searchResult?.blogs?.map`

File: src/components/Layouts/components/Search/Search.tsx [hunk 9]
```diff
@@ -95,7 +94,7 @@ const Search = () => {
...
                       <SearchItem title={blog.title} thumnails={blog.banner} key={blog._id} blogId={blog._id} />
```
Comment (nash-nguyen) [inline src/components/Layouts/components/Search/Search.tsx]: Check null `blog` object

File: src/pages/Home/Home.tsx [hunk 10]
```diff
@@ -39,11 +47,31 @@ const Home = () => {
...
+    return () => {
+      socket?.disconnect()
+    }
+  }, [])
+
+  useEffect(() => {
+    if (userInfo) {
+      socket?.emit('newUser', userInfo._id)
+    }
+  }, [socket, userInfo])
```
Comment (nash-nguyen) [inline src/pages/Home/Home.tsx]: `}, [socket, userInfo])`

to 

`}, [socket, userInfo?._id])`

**Review by hoangdaheo — APPROVED** [review]

**Pull request merged**